tests/test.c

```diff
@@ -660,6 +660,22 @@ static int p_argon2i_overlap()
     return status;
 }
 
+// Tests that the shared key and secret key buffers of crypto_key_exchange can overlap.
+static int p_key_exchange_overlap()
+{
+    int status = 0;
+    FOR (i, 0, 5) {
+        u8 buf[32];
+        RANDOM_INPUT(sk, 32);
+        RANDOM_INPUT(pk, 32);
+        crypto_key_exchange(buf, sk, pk);
+        crypto_key_exchange(sk, sk, pk);
+        status |= memcmp(buf, sk, 32);
+    }
```

Comment:
I agree with @fscoto on this one, ideally we should have a bigger buffer and test every possible overlap (62 tests total?). While partial overlap don't make sense, making sure they are possible makes sure the code resist some overly clever uses.

So:

    FOR (i, 0, 64) {
        u8 overlapping[96];
        u8 separate[32];
        RANDOM_INPUT(sk, 32);
        RANDOM_INPUT(pk, 32);
        memcmp(overlapping+32, sk, 32);
        crypto_key_exchange(separate, sk, pk);
        crypto_key_exchange(overlapping + i, overlapping + 32, pk);
        status |= memcmp(separate, overlapping, 32);
    }

(Not tested, not checked the consistency with the rest of the test suite, but you get the idea.)